### PR TITLE
Delete all

### LINKED ISSUE #525: Add an API for blowing away all data
It would help for development purposes, especially since now we are persisting data on the host machine for local deployment.

### PR DESCRIPTION
**Pull request by jdoliner**
Fixes #525

### CODE REVIEW

**Comment by jdoliner** [comment]
One issue here is that there isn't a good way to tests this. Calling `DeleteAll` from the integration tests will definitely interfere with the other tests passing.

**Comment by derekchiang** [comment]
@jdoliner can't you just not do `t.Parallel()` in the test, so that the test doesn't interfere with other tests?

File: src/server/cmd/pachctl/cmd/cmd.go [hunk 1]
```diff
@@ -66,7 +68,29 @@ Envronment variables:
 			return writer.Flush()
 		}),
 	}
+	deleteAll := &cobra.Command{
+		Use:   "delete-all",
+		Short: "Delete everything.",
+		Long:  "Delete everything.",
```
Comment (derekchiang) [inline src/server/cmd/pachctl/cmd/cmd.go]: Can you explain what "everything" means in the `Long` help string?

File: src/server/cmd/pachctl/cmd/cmd.go [hunk 2]
```diff
@@ -66,7 +68,29 @@ Envronment variables:
...
+		Run: pkgcobra.RunFixedArgs(0, func(args []string) error {
+			client, err := client.NewFromAddress(address)
+			if err != nil {
+				return err
+			}
+			fmt.Printf("Are you sure you want to delete everything? yN\n")
```
Comment (derekchiang) [inline src/server/cmd/pachctl/cmd/cmd.go]: Maybe briefly state what "everything" means here as well?

File: src/server/pfs/drive/driver.go [hunk 3]
```diff
@@ -725,6 +726,36 @@ func (d *driver) deleteFile(file *pfs.File, shard uint64, unsafe bool, handle st
...
+
+func (s byProvenance) Len() int {
+	return len(s)
+}
+func (s byProvenance) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+func (s byProvenance) Less(i, j int) bool {
+	return len(s[i].Provenance) < len(s[j].Provenance)
+}
```
Comment (derekchiang) [inline src/server/pfs/drive/driver.go]: Smart

File: src/server/pfs/drive/driver.go [hunk 4]
```diff
@@ -725,6 +726,36 @@ func (d *driver) deleteFile(file *pfs.File, shard uint64, unsafe bool, handle st
...
+	// We want to make sure we delete repos before their provenance.
+	// Provenance has a nice invariant:
+	// A in Provenance(B) => len(Provenance(A)) < len(Provenance(B))
+	// Since Provenance is transitive
+	// Thus when we sort by length we guarantee that A will have a lower index than B
+	// Because we want to make sure we delete the higher indexes first we traverse in reverse
+	sort.Sort(byProvenance(repoInfos))
+	for i := range repoInfos {
+		d.DeleteRepo(repoInfos[len(repoInfos)-i-1].Repo, shards)
+	}
```
Comment (derekchiang) [inline src/server/pfs/drive/driver.go]: We are ignoring errors from DeleteRepo?

File: src/server/pps/server/api_server.go [hunk 5]
```diff
@@ -889,6 +889,18 @@ func (a *apiServer) DeletePipeline(ctx context.Context, request *ppsclient.Delet
 	return google_protobuf.EmptyInstance, nil
 }
 
+func (a *apiServer) DeleteAll(ctx context.Context, request *google_protobuf.Empty) (response *google_protobuf.Empty, retErr error) {
+	defer func(start time.Time) { a.Log(request, response, retErr, time.Since(start)) }(time.Now())
+	persistClient, err := a.getPersistClient()
+	if err != nil {
+		return nil, err
+	}
+	if _, err := persistClient.DeleteAll(ctx, request); err != nil {
```
Comment (derekchiang) [inline src/server/pps/server/api_server.go]: Not sure if this is quite enough.  What about the goroutines that are running `runPipeline`?  Probably wanna cancel them using `cancelFuncs`.

File: src/server/cmd/pachctl/cmd/cmd.go (hunk 1 above)
Comment (jdoliner) [inline src/server/cmd/pachctl/cmd/cmd.go]: Surely.

File: src/server/cmd/pachctl/cmd/cmd.go (hunk 2 above)
Comment (jdoliner) [inline src/server/cmd/pachctl/cmd/cmd.go]: Done

File: src/server/pfs/drive/driver.go (hunk 3 above)
Comment (jdoliner) [inline src/server/pfs/drive/driver.go]: :D

File: src/server/pfs/drive/driver.go (hunk 4 above)
Comment (jdoliner) [inline src/server/pfs/drive/driver.go]: Lol, we should not be.

Comment (jdoliner) [inline src/server/pfs/drive/driver.go]: Should be fixed now.

File: src/server/pps/server/api_server.go (hunk 5 above)
Comment (jdoliner) [inline src/server/pps/server/api_server.go]: Great point.

Comment (jdoliner) [inline src/server/pps/server/api_server.go]: Alright, now it calls `deletePipeline` just like `DeletePipeline` does.

**Comment by jdoliner** [comment]
This now has a test as well. Still waiting to see if it works on travis... I have my doubts.

**Comment by jdoliner** [comment]
Uh oh, I think I somehow completely blew away this pull request with git... that's less than ideal.

**Pull request merged**